### PR TITLE
feat(autoexport): change WithFallback options signatures

### LINKED ISSUE #4877: autoexport: Change WithFallback options signatures
### Problem Statement

The current definition of fallback options is as following:

```go
func WithFallbackMetricReader(exporter metric.Reader) MetricOption
```

The problem is that we need to create an instance of a metric reader or span exporter and it may not even be needed/used (it may even leak resources).

### Proposed Solution

Change the signature to accept a factory function e.g.

```go
func WithFallbackMetricReader(factory func(context.Context) (metric.Reader, error)) MetricOption
```

### PR DESCRIPTION
**Pull request by hcelaloner**
Fixes #4877.

Changes the method signatures as requested in the issue.

### CODE REVIEW

**Review by MrAlias — COMMENTED** [review]
Please add a changelog entry for this breaking change.

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add client metric support to `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`. (#4707)
 - Add peer attributes to spans recorded by `NewClientHandler`, `NewServerHandler` in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc`. (#4873)
 
+### Changed
+
+- Change the signature of the fallback options in  `go.opentelemetry.io/contrib/exporters/autoexport` to accept a factory function instead. `WithFallbackMetricReader(exporter metric.Reader) MetricOption` method is now replaced with `func WithFallbackMetricReader(metricReaderFactory func(ctx context.Context) (metric.Reader, error)) MetricOption` method. `WithFallbackSpanExporter(exporter trace.SpanExporter) SpanOption` method is now replaced with `WithFallbackSpanExporter(spanExporterFactory func(ctx context.Context) (trace.SpanExporter, error)) SpanOption` method. (#4891)
```
Comment (MrAlias) [inline CHANGELOG.md:19]:
```suggestion
- The fallback options in  `go.opentelemetry.io/contrib/exporters/autoexport` now accept factory functions. (#4891)
  - `WithFallbackMetricReader(metric.Reader) MetricOption` is replaced with `func WithFallbackMetricReader(func(ctx context.Context) (metric.Reader, error)) MetricOption`.
  - `WithFallbackSpanExporter(trace.SpanExporter) SpanOption` is replaced with `WithFallbackSpanExporter(func(ctx context.Context) (trace.SpanExporter, error)) SpanOption`.
```

**Review by MrAlias — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

File: exporters/autoexport/signal.go [hunk 2]
```diff
@@ -51,8 +51,8 @@ func (s signal[T]) create(ctx context.Context, opts ...option[T]) (T, error) {
 }
 
 type config[T any] struct {
-	hasFallback bool
-	fallback    T
+	hasFallback     bool
```
Comment (pellared) [inline exporters/autoexport/signal.go:54]: Can we get rid of `hasFallback` and simply use `if fallbackFactory != nil` instead?

Comment (hcelaloner) [reply exporters/autoexport/signal.go:54]: Yes, you are right it seems we can get rid of "hasFallback" 👍

**Review by pellared — APPROVED** [review]

**Pull request merged**